src/cmd-cave.c

```diff
@@ -1340,6 +1339,264 @@ void do_cmd_run(struct command *cmd)
+		return;
+	}
+	
+
+	int min_dist = 9999;
+	/* Get location */
+	for (int y = 0; y < cave->height; y++) {
+		for (int x = 0; x < cave->width; x++) {
+			struct loc grid = loc(x, y);
+			
+			if (loc_eq(grid, player->grid)) continue;
+
+			if (square_isoccupied(cave, grid)) {
+				int m_idx = square(cave, grid)->mon;
+				struct monster *mon = cave_monster(cave, m_idx);
+				if (monster_is_obvious(mon)) {
+					visible_monster_count++;
+					break;
+				}
+			}
+			if (square_isknown(cave, grid) && square_isupstairs(cave, grid)) {
+				int d = distance(grid, player->grid);
+				if (d < min_dist) {
+					target_grid = grid;
+					min_dist = d;
+				}
+			}
+		}
+	}
+
+	if (visible_monster_count > 0) {
+		msg("Something is here.");
+		return;
+	}
+
+	if (min_dist < 9999 && find_path(target_grid)) {
+		player->upkeep->running = 1000;
+		/* Calculate torch radius */
+		player->upkeep->update |= (PU_TORCH);
+		player->upkeep->running_withpathfind = true;
+		run_step(0);
+		return;
+	}
+
+	msg("No known path to upstairs.");
+}
+
+/**
+ * Start exploring.
+ *
+ * Note that exploring while confused is not allowed.
+ */
+void do_cmd_explore(struct command *cmd)
+{
+	struct loc target_grid;
+	bool visible_monster = false;
+	/* cancel if confused */
+	if (player->timed[TMD_CONFUSED]) {
+		msg("You cannot explore while confused.");
+	   	return;
+	}
+
+
+	/* If we're in a web, deal with that */
+	if (square_iswebbed(cave, player->grid)) {
+		/* Clear the web, finish turn */
+		msg("You clear the web.");
+		square_destroy_trap(cave, player->grid);
+		player->upkeep->energy_use = z_info->move_energy;
+		return;
+	}
+	
+
+	int min_dist = 9999;
+	/* Get location */
+	for (int y = 0; y < cave->height && !visible_monster; y++) {
+		for (int x = 0; x < cave->width; x++) {
+			struct loc grid = loc(x, y);
+			
+			if (loc_eq(grid, player->grid)) continue;
+
+			if (square_isoccupied(cave, grid)) {
+				int m_idx = square(cave, grid)->mon;
+				struct monster *mon = cave_monster(cave, m_idx);
+				if (monster_is_obvious(mon)) {
+					visible_monster = true;
+					break; /* only breaks the inner loop */
+				}
+			}
+			if (square_isnotknown(cave, grid) && square_ispassable(cave, grid)) {
+				int d = distance(grid, player->grid);
+				if (d < min_dist) {
+					target_grid = grid;
+					min_dist = d;
+				}
+			}
+		}
+	}
+
+	if (visible_monster) {
+		msg("Something is here.");
+		return;
+	}
+
+	if (min_dist < 9999 && find_path(target_grid)) {
+		player->upkeep->running = 1000;
+		/* Calculate torch radius */
+		player->upkeep->update |= (PU_TORCH);
+		player->upkeep->running_withpathfind = true;
+		run_step(0);
+		return;
+	}
+
+	/* 
+	 * No unexplored location found, walk to a door/rubble grid position,
+	 * that has an explored and unexplored side.
+	 */
+	min_dist = 9999;
+	for (int y = 0; y < cave->height; y++) {
+		for (int x = 0; x < cave->width; x++) {
+			struct loc grid = loc(x, y);
+
+			if (!square_iscloseddoor(cave, grid) && !square_isrubble(cave, grid))
```

Comment:
From here to the end of the loop also leaks information since it can probe the terrain type of an unknown location.  A version that wouldn't leak information would be:

```
if (!square_isknown(cave, grid) && !square_iscloseddoor(cave, grid) && !square_isrubble(cave, grid)) continue;

if (count_neighbors(NULL, cave, grid, square_isnotknown, false) > 0) {
    int d = distance(grid, player->grid);

    if (d < min_dist) {
        target_grid = d;
        min_dist = d;
    }
}
```

In the information-leaking version, you use !square_in_bounds(cave, neighbor) || !square_isfloor(cave, neighbor) to reject uninteresting locations next to the door/rubble.  Shouldn't that be !square_in_bounds(cave, neighbor) || !(square_ispassable(cave, heighbor) || square_iscloseddoor(cave, neighbor) || square_isrubble(cave, neighbor)) ?

Reply:
I am not sure about the last part, simply because I am not sure about the semantics of "ispassable", but my guess would be that you'd only bring this up if "ispassable" does not include "isrubble" and "iscloseddoor", which would make sense since these are only potentially passable.

Reply:
I was thinking of cases where the rubble/door was next to a unknown staircase (passable but not a floor), unknown rubble (not a floor and depending on the type of rubble, not passable), or an unknown closed door (not a floor and not passable); on the explored side of the rubble/door you are right that you'd want a passable location as the destination for the run.

The version I proposed for handling door/rubble without leaking information does not get a target_grid that is passable, but you caught that in your revision.